Curve448 primitives

mattcaswell opened the pull request:
This adds all the primitives required for Curve448 support (X448 and Ed448).

This is based heavily on original work by Mike Hamburg (thanks @bitwiseshiftleft) and contributed under CLA.

This is code that originally appeared in #4829, but I've pulled out just the primitives (i.e. none of the EVP changes). The EVP layer will come in a different PR once this is in. I've also stripped out all commits that were updating the temporary Makefile I had in #4829 , so that doesn't appear here.

At the moment only the 32-bit reference code is used. There are a number of files for other architectures that I have imported but done very little else with at this stage. Again later PRs can look at integrating the platform specific code in.

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING

Other than that, provide a description above this comment if there isn't one already

If this fixes a github issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [x] tests are added or updated

stevelinsell commented on crypto/ec/curve448/arch_arm_32/f_impl.h:
```diff
@@ -0,0 +1,59 @@
+
+void gf_add_RAW(gf out, const gf a, const gf b)
+{
+    for (unsigned int i = 0; i < sizeof(*out) / sizeof(uint32xn_t); i++) {
```
Minor nit, declare i outside of the for loop for pre C99 compatibility. Same in other places.

mattcaswell replied:
Right - likely there are quite a few issues in the non 32-bit reference code (basically anything in one of the arch* directories that isn't arch_32). As I said in my intro to this, none of that code is actually used right now, and I simply imported it from Mike Hamburg's original contribution so we have the ability to hook it up later.

stevelinsell replied:
Yep, sorry missed the 'arm' bit when reviewing this header, should have spotted it was code outside the review of just arch_32.

mattcaswell commented:
Added a commit to fix an AppVeyor failure. Ping @openssl/committers - still looking for a reviewer for this.

bernd-edlinger commented on crypto/ec/curve448/arch_32/f_impl.c:
```diff
@@ -0,0 +1,102 @@
+# define REPEAT8(_x) _x _x _x _x _x _x _x _x
+# define FOR_LIMB(_i,_start,_end,_x) do { _i=_start; REPEAT8( if (_i<_end) { _x; } _i++;) } while (0)
+#else
+# define FOR_LIMB(_i,_start,_end,_x) do { for (_i=_start; _i<_end; _i++) _x; } while (0)
```
I'd like to see macro parameters in parentheses

mattcaswell replied:
I replaced this macro with your version.

bernd-edlinger commented on crypto/ec/curve448/arch_32/f_impl.c:
```diff
@@ -0,0 +1,102 @@
+             accum2 = 0;
+             FOR_LIMB(i, j + 1, 8, {
+                      accum0 -=
+                      widemul(a[8 + j - i], b[i]);
```
Indentation looks wrong

mattcaswell replied:
Fixed

bernd-edlinger commented on crypto/ec/curve448/arch_32/f_impl.c:
```diff
@@ -0,0 +1,102 @@
+
+    assert(b < 1 << 28);
+
+    FOR_LIMB(i, 0, 8, {
```
I'm not sure I like this new for loop syntax...

mattcaswell replied:
It is as written in the original implementation by Mike Hamburg. We could choose to rewrite it anyway we like. Do you have any ideas about how else to do it? One of the problems of putting code blocks in a macro like this is that it makes it difficult to run through a debugger.

bernd-edlinger replied:
Yes, I just saw the ugliness so far, but debugging must be near impossible this way.
I think this "FOR_LIMB(i, 0, 8, {" does not add much value over "for (i=0; i<8; i++) {"

mattcaswell replied:
Well I think the justification for it is performance, i.e. we normally use this version of the macro:

````
# define REPEAT8(_x) _x _x _x _x _x _x _x _x
# define FOR_LIMB(_i,_start,_end,_x) do { _i=_start; REPEAT8( if (_i<_end) { _x; } _i++;) } while (0)
````

Which is written specifically to *avoid* a loop. The original code has been significantly optimised by Mike. During importing the code I was trying to avoid doing anything that was going to break that optimisation.

bernd-edlinger replied:
Hmm....

This looks more like a duff's device

`# define FOR_LIMB(_i, _start, _end, _x) do { (_i) = (_start); switch((_end)-(_start)) { case 8: _x; (_i)++; case 7:  _x;  (_i)++;  .... case 1: _x; (_i)++; } } while(0)`

mattcaswell replied:
I used your version of the macro.

bernd-edlinger commented on crypto/ec/curve448/arch_32/f_impl.h:
```diff
@@ -0,0 +1,56 @@
+{
+    unsigned int i;
+
+    for (i = 0; i < sizeof(*out) / sizeof(out->limb[0]); i++) {
```
use OSSL_NELEM here?

mattcaswell replied:
Actually I just replaced this with `NLIMBS` which seems more appropriate.

bernd-edlinger commented on crypto/ec/curve448/arch_neon/f_impl.h:
```diff
@@ -0,0 +1,65 @@
+void gf_add_RAW(gf out, const gf a, const gf b)
+{
+    for (unsigned int i = 0; i < sizeof(*out) / sizeof(uint32xn_t); i++) {
+        ((uint32xn_t *) out)[i] =
```
this looks like a strict-aliasing violation

mattcaswell replied:
The comments in the architecture specific files except anything inside arch_32 have barely been touched by my at all. They are not used or compiled. I only wanted to import them because I thought we would probably integrate them at a later stage. If that is considered a blocker for review I could remove these files completely (but then we would probably add them back in again at some later time).

bernd-edlinger replied:
yes, that might be better...

mattcaswell replied:
I removed this file. It still gets imported in the original commit, and then gets deleted again. This means we can just resurrect the code easily without having to worry about the code's provenance/CLAs etc.

bernd-edlinger commented on crypto/ec/curve448/arch_ref64/arch_intrinsics.h:
```diff
@@ -0,0 +1,31 @@
+
+# define ARCH_WORD_BITS 64
+
+static __inline__ __attribute((always_inline, unused))
```
the keywords always_inline, unused etc, can also be written as __always_inline__, __unused__

mattcaswell replied:
I deleted this file

bernd-edlinger commented on crypto/ec/curve448/arch_ref64/arch_intrinsics.h:
```diff
@@ -0,0 +1,31 @@
+static __inline__ __attribute((always_inline, unused))
+uint64_t word_is_zero(uint64_t a)
+{
+    /* let's hope the compiler isn't clever enough to optimize this. */
```
Don't we have a constant_time-function for that?

kroeckx replied:
Yes, and I've already pointed that out in the previous pull request

mattcaswell replied:
I deleted this file, although this function appears elsewhere so I guess the same comment applies. As @kroeckx mentioned he previously raised this. My response at that time was:

> The types aren't quite the same, so this is probably a non-trivial exercise.

i.e. it is a non-trivial exercise to merge them (really they are different functions because of the different types)

bernd-edlinger commented on crypto/ec/curve448/constant_time.h:
```diff
@@ -0,0 +1,337 @@
+ * Note that the output is not __restrict__, but if it overlaps either
+ * input, it must be equal and not partially overlap.
+ */
+static ossl_inline void constant_time_select(void *a_,
```
at least this function might collide with one of our constant time functions

mattcaswell replied:
I renamed it

bernd-edlinger commented on crypto/ec/curve448/arch_32/f_impl.c:
```diff
@@ -0,0 +1,102 @@
+
+#include "field.h"
+
+#if (defined(__OPTIMIZE__) && !defined(__OPTIMIZE_SIZE__) && !defined(I_HATE_UNROLLED_LOOPS)) \
```
instead of "I_HATE.." we should use "OPENSSL_SMALL_FOOTPRINT"

mattcaswell replied:
Actually I deleted these and just had the unrolled loop version. I'm not sure why we would ever not want to do that.

mattcaswell commented:
Updated commits pushed addressing comments so far.

bernd-edlinger commented on crypto/ec/curve448/arch_32/f_impl.c:
```diff
@@ -0,0 +1,111 @@
+#include "field.h"
+
+/* Use an unrolled loop for performance */
+#define FOR_LIMB(_i, _start, _end, _x) \
```
Thanks, looks good, I would however like to use OPENSSL_SMALL_FOOTPRINT
to Switch to the non-inlined variant.

mattcaswell replied:
Ok. Done. I moved this define to curve448utils.h so that it could be picked up by f_generic.c too.

bernd-edlinger commented on crypto/ec/curve448/constant_time.h:
```diff
@@ -0,0 +1,337 @@
+        } else {
+            /* aligned */
+            big_register_t xor = *((big_register_t *) (&a[k]))
+                                 ^ *((big_register_t *) (&b[k]));
```
Hmm....

big_register_t is likely not the effective type of the object pointed to with a, and b.
Thus this seems to be in violation of the strict-aliasing rules.

mattcaswell replied:
Hmmm. Nasty - could be tricky to fix. I decided to bite the bullet and rip out all the curve448 specific constant time stuff and use our standard calls. It may result in some performance degradation, but I guess we worry about that later.

bernd-edlinger commented on crypto/ec/curve448/constant_time.h:
```diff
@@ -0,0 +1,337 @@
+# endif
+            } else {
+                /* aligned */
+                word_t xor = *((word_t *) (&a[k])) ^ *((word_t *) (&b[k]));
```
and word_t is yet another type.

mattcaswell replied:
As above.

bernd-edlinger commented on crypto/ec/curve448/curve448utils.h:
```diff
@@ -0,0 +1,86 @@
+ */
+# ifndef C448_WORD_BITS
+#  if (defined(__ILP64__) || defined(__amd64__) || defined(__x86_64__) \
+      || (((__UINT_FAST32_MAX__)>>30)>>30))
```
any idea why we are shifting twice?
why 60 bits, and not 63?

kroeckx replied:
I'm guessing to avoid a warning if it's 32 bit?

mattcaswell replied:
Don't know. Maybe @bitwiseshiftleft can enlighten us.

dot-asm replied:
This is not working for non-GNU (or non-GNU-alikes) or elderly GNU compilers. Well, not this #if specifically, but what it enables, __int128 few lines below. I'm not really clear on which type is officially supported as builtin in long term, [and there were some changes,] but for poly1305 I've chosen __int128 alone, i.e. `unsigned __int128` and not __uint128_t. Well, it doesn't mean it's **the one**, but it served well so far. As for pre-processor condition poly1305 uses
```
# if !defined(PEDANTIC) && \
     (defined(__SIZEOF_INT128__) && __SIZEOF_INT128__==16) && \
     (defined(__SIZEOF_LONG__) && __SIZEOF_LONG__==8)
```
with rationale that it filters out non-GNU (and non-alikes) as well as elderly GNU.

Just in case, "not working" means "actually tried and confirmed".

bernd-edlinger replied:
Either use @dot-asm's suggestion for the preprocessor condition or
just take this part out for the moment, C448_WORD_BITS = 32 should always work.

bernd-edlinger replied:
I think we change this condition to be simply:
```
# if defined(__SIZEOF_INT128__) && __SIZEOF_INT128__==16
```

what do you think?

mattcaswell replied:
Works for me. Done.

bernd-edlinger commented on crypto/ec/curve448/curve448utils.h:
```diff
@@ -0,0 +1,86 @@
+    return (c448_error_t) x;
+}
+
+#ifdef __cplusplus
```
why use extern "C" binding?

kroeckx replied:
Since this is an internal only header, I see little point. We do seem to be doing it for some of the other internal headers, but clearly not all.

mattcaswell replied:
Remnants from the imported code. I deleted all occurrences of this.

bernd-edlinger commented on crypto/ec/curve448/eddsa.c:
```diff
@@ -0,0 +1,363 @@
+        uint8_t challenge[2 * EDDSA_448_PRIVATE_BYTES];
+
+        if (hashctx == NULL
+            || !hash_init_with_dom(hashctx, prehashed, 0, context, context_len)
```
we want extra indenting here.

mattcaswell replied:
Fixed

bernd-edlinger commented on crypto/ec/curve448/f_generic.c:
```diff
@@ -0,0 +1,200 @@
+    unsigned int i;
+    mask_t succ;
+
+    UNROLL for (i = 0; i < NLIMBS; i++) {
```
yet another differnet way of loop unrolling.
this one is limited to clang, so I have to object :)

bernd-edlinger replied:
how about this:

```
#define FOR_LIMB(_i, _start, _end, _x) \
    do { \
      (_i) = (_start); \
      do { \ 
        switch((_end) - (_i)) { \
        default: \
        case 8: _x; (_i)++; /* fall through */ \
        case 7: _x; (_i)++; /* fall through */ \
        case 6: _x; (_i)++; /* fall through */ \
        case 5: _x; (_i)++; /* fall through */ \
        case 4: _x; (_i)++; /* fall through */ \
        case 3: _x; (_i)++; /* fall through */ \
        case 2: _x; (_i)++; /* fall through */ \
        case 1: _x; (_i)++; continue; \
        case 0: break; \
        } \
     } while(0); \
    } while(0)
```
could translate
`UNROLL while (fill < LIMB_PLACE_VALUE(LIMBPERM(i)) && j < nbytes) {`
to
`FOR_LIMB(j, j, nbytes, { if (fill < LIMB_PLACE_VALUE(LIMBPERM(i))) break; ...`

bernd-edlinger replied:
Aehm, correction:

```
        case 1: _x; (_i)++; continue; \
        case 0: break; \
        } \
     break; \
     } while(1); \
   } while(0)
```

mattcaswell replied:
Done

bernd-edlinger commented on crypto/ec/curve448/arch_32/f_impl.h:
```diff
@@ -0,0 +1,56 @@
+{
+    unsigned int i;
+
+    for (i = 0; i < sizeof(*out) / sizeof(out->limb[0]); i++) {
```
use NLIMBS

mattcaswell replied:
Done

bernd-edlinger commented on crypto/ec/curve448/arch_32/f_impl.h:
```diff
@@ -0,0 +1,56 @@
+    unsigned int i;
+    uint32_t co1 = ((1 << 28) - 1) * amt, co2 = co1 - amt;
+
+    for (i = 0; i < sizeof(*a) / sizeof(a->limb[0]); i++)
```
ditto

mattcaswell replied:
Done

bernd-edlinger commented on crypto/ec/curve448/arch_32/f_impl.h:
```diff
@@ -0,0 +1,56 @@
+    unsigned int i;
+
+    a->limb[8] += tmp;
+    for (i = 15; i > 0; i--)
```
Loop terminates with unsigned underflow, maybe make i signed?

mattcaswell replied:
I don't see an underflow? Loop terminates at i == 0?

bernd-edlinger replied:
yes, forget it.

kroeckx commented on crypto/ec/curve448/arch_32/arch_intrinsics.h:
```diff
@@ -0,0 +1,29 @@
+static ossl_inline uint32_t word_is_zero(uint32_t a)
+{
+    /* let's hope the compiler isn't clever enough to optimize this. */
+    return (((uint64_t)a) - 1) >> 32;
```
I can't find the previous comment about this anymore, probably because it was on a file that's now deleted.

It's true that we currently don't have a version that takes a uint32_t (and uint64_t), but we've already added an uint64_t of the msb function, it shouldn't be that hard to add a few more functions.

mattcaswell replied:
Now using our own standard constant time code for everything constant time related.

mattcaswell commented:
New commits pushed addressing all comments.

bernd-edlinger commented on crypto/ec/curve448/curve448utils.h:
```diff
@@ -0,0 +1,107 @@
+# define FOR_LIMB(_i,_start,_end,_x) \
+    do { \
+        for ((_i) = (_start); (_i) < (_end); (_i)++) \
+            _x; \
```
maybe add some braces around _x here?

mattcaswell replied:
Done. I also did the same to the unrolled version

bernd-edlinger commented on crypto/ec/curve448/eddsa.c:
```diff
@@ -0,0 +1,363 @@
+        return C448_FAILURE;
+
+    if (!EVP_DigestInit_ex(hashctx, EVP_shake256(), NULL)
+        || !EVP_DigestUpdate(hashctx, in, inlen)
```
indentation.

mattcaswell replied:
Fixed

bernd-edlinger commented on crypto/ec/curve448/eddsa.c:
```diff
@@ -0,0 +1,363 @@
+        return C448_FAILURE;
+
+    if (!EVP_DigestInit_ex(hashctx, EVP_shake256(), NULL)
+        || !EVP_DigestUpdate(hashctx, dom_s, strlen(dom_s))
```
indentation.

mattcaswell replied:
Fixed

bernd-edlinger commented on crypto/ec/curve448/eddsa.c:
```diff
@@ -0,0 +1,363 @@
+
+        /* Hash to create the nonce */
+        if (!hash_init_with_dom(hashctx, prehashed, 0, context, context_len)
+            || !EVP_DigestUpdate(hashctx, expanded + EDDSA_448_PRIVATE_BYTES,
```
indentation.

mattcaswell replied:
Fixed

bernd-edlinger commented on crypto/ec/curve448/f_generic.c:
```diff
@@ -0,0 +1,203 @@
+    gf_sub(c, a, b);
+    gf_strong_reduce(c);
+
+    for (i = 0; i < NLIMBS; i++) {
```
do we need braces here?

mattcaswell replied:
No. Removed.

mattcaswell commented:
New commit pushed addressing comments. I also tweaked the `constant_time_lookup()` function to add an explicit cast of the mask to avoid an AppVeyor failure.

bernd-edlinger commented on crypto/ec/curve448/word.h:
```diff
@@ -0,0 +1,171 @@
+
+static ossl_inline void ignore_result(c448_bool_t boo)
+{
+    (void)boo;
```
what is the purpose here?

mattcaswell replied:
There are a couple of places where we are explicitly deciding to ignore the result of a function call. This just makes it clearer that it is intentional. Since there are now only 2 of these I could delete this function altogether and just cast to void directly...but I kind of thought this was ok.

bernd-edlinger replied:
yes, agreed.  I was just unsure if this was supposed to force the compiler to not optimize the
ignored expression away.

bernd-edlinger commented on crypto/ec/curve448/curve448utils.h:
```diff
@@ -0,0 +1,107 @@
+# endif
+
+/* C448_TRUE = -1 so that C448_TRUE & x = x */
+static const c448_bool_t C448_TRUE = 0 - (c448_bool_t)1;
```
why is this not simply a define?

mattcaswell replied:
It is now.

bernd-edlinger commented on crypto/ec/curve448/eddsa.c:
```diff
@@ -0,0 +1,364 @@
+                                     expanded + EDDSA_448_PRIVATE_BYTES,
+                                     EDDSA_448_PRIVATE_BYTES)
+                || !EVP_DigestUpdate(hashctx, message, message_len)) {
+                OPENSSL_cleanse(expanded, sizeof(expanded));
```
indentation to deep here

bernd-edlinger commented on crypto/ec/curve448/eddsa.c:
```diff
@@ -0,0 +1,364 @@
+        curve448_point_t p;
+
+        curve448_scalar_halve(nonce_scalar_2, nonce_scalar);
+        for (c = 2; c < C448_EDDSA_ENCODE_RATIO; c <<= 1) {
```
unnecessary braces

bernd-edlinger commented on crypto/ec/curve448/eddsa.c:
```diff
@@ -0,0 +1,364 @@
+
+        /* Compute the challenge */
+        if (!hash_init_with_dom(hashctx, prehashed, 0, context, context_len)
+            || !EVP_DigestUpdate(hashctx, nonce_point, sizeof(nonce_point))
```
indentation

bernd-edlinger commented on crypto/ec/curve448/eddsa.c:
```diff
@@ -0,0 +1,364 @@
+{
+    c448_error_t ret;
+
+    ret = c448_ed448_verify(signature, pubkey, hash, 64, 1, context,
```
maybe simply `return c448_ed448_verify(...);`

bernd-edlinger commented on crypto/ec/curve448/f_generic.c:
```diff
@@ -0,0 +1,202 @@
+/* Return high bit of x = low bit of 2x mod p */
+mask_t gf_hibit(const gf x)
+{
+    gf y;
```
empty line please.

bernd-edlinger commented on crypto/ec/curve448/f_generic.c:
```diff
@@ -0,0 +1,202 @@
+/* Return high bit of x = low bit of 2x mod p */
+mask_t gf_lobit(const gf x)
+{
+    gf y;
```
and here

bernd-edlinger commented on crypto/ec/curve448/f_generic.c:
```diff
@@ -0,0 +1,202 @@
+
+mask_t gf_isr(gf a, const gf x)
+{
+    gf L0, L1, L2;
```
empty line here

bernd-edlinger commented on crypto/ec/curve448/field.h:
```diff
@@ -0,0 +1,172 @@
+# define SER_BYTES 56
+
+# if defined(__GNUC__) || defined(__clang__)
+#  define INLINE_UNUSED __inline__ __attribute__((unused,always_inline))
```
bare word like `unused` and `always_inline` could collide with defines
so `__unused__` and `__always_inline__` should be more portable.
OTOH this is an internal header file, so that is not a must.

mattcaswell replied:
Fixed

bernd-edlinger commented on crypto/ec/curve448/ed448.h:
```diff
@@ -0,0 +1,199 @@
+ *
+ * For Ed25519, it is unsafe to use the same key for both prehashed and
+ * non-prehashed messages, at least without some very careful protocol-level
+ * disambiguation.  For Ed448 it is safe.  The C++ wrapper is designed to make
```
this refers to a C++ wrapper, but we won't have that, right?

mattcaswell replied:
Right. I removed this reference.

bernd-edlinger replied:
did you?

bernd-edlinger commented on crypto/ec/curve448/field.h:
```diff
@@ -0,0 +1,172 @@
+# if defined(__GNUC__) || defined(__clang__)
+#  define INLINE_UNUSED __inline__ __attribute__((unused,always_inline))
+#  define RESTRICT __restrict__
+#  define ALIGNED __attribute__((aligned(32)))
```
this requests 32 _byte_ alignment. The alignment is not measured in bits here.
Was that the intention?
That is more aligned that what we usually get from malloc.

mattcaswell replied:
As per comment from Mike:
"32-byte alignment is required for avx2 operations.  On x86 at least, the library may be using avx2 even if the caller isn’t compiled with -mavx2.  We could instead change it so all the vector ops assume unaligned."

bernd-edlinger commented on crypto/ec/curve448/word.h:
```diff
@@ -0,0 +1,171 @@
+ * Originally written by Mike Hamburg
+ */
+
+#ifndef __WORD_H__
```
not sure if we should use double underscore
i.e. reserved identifier for header inclusion macros.
I see other headers use a HEADER_ prefix for this.

bernd-edlinger commented on crypto/ec/curve448/word.h:
```diff
@@ -0,0 +1,171 @@
+/* PERF: vectorize vs unroll */
+# ifdef __clang__
+#  if 100*__clang_major__ + __clang_minor__ > 305
+#   define UNROLL _Pragma("clang loop unroll(full)")
```
could we remove this please.

mattcaswell replied:
Done

bernd-edlinger commented on test/curve448_internal_test.c:
```diff
@@ -0,0 +1,702 @@
+    static uint8_t hashout[64];
+
+    if (!EVP_DigestInit_ex(hashctx, EVP_shake256(), NULL)
+        || !EVP_DigestUpdate(hashctx, msg, msglen)
```
indentation wrong.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+
+extern const gf curve448_precomputed_base_as_fe[];
+const curve448_precomputed_s *curve448_precomputed_base =
+    (const curve448_precomputed_s *)&curve448_precomputed_base_as_fe;
```
aliasing violation here.

mattcaswell replied:
Fixed this by changing the type of the precomputed tables so we can avoid this cast.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+    ret = gf_isr(t2, t1);       /* +-1/sqrt(o^2) = +-1/o */
+    (void)ret;
+    if (assert_nonzero)
+        assert(ret);
```
IIRC, some OMC members don't like adding new asserts

mattcaswell replied:
As discussed yesterday on openssl-project we prefer ossl_assert over assert and don't want to see OPENSSL_assert() at all in libcrypto or libssl. In this case (and all the others you mention) I think use of simple "assert" is actually justified. All of these instances are either in void functions or functions that do have a return value, but don't return errors. I don't think it is worth the effort of changing them so that they do - so just plain assert should be fine.

bernd-edlinger replied:
ok, fine.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+static void sub_niels_from_pt(curve448_point_t d, const niels_t e,
+                              int before_double)
+{
+    gf a, b, c;
```
empty line missing.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+    mask_t succ;
+
+    /* equality mod 2-torsion compares x/y */
+    gf a, b;
```
move this to the variable section

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+{
+    mask_t out;
+
+    gf a, b, c;
```
here as well

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+    const unsigned int n = COMBS_N, t = COMBS_T, s = COMBS_S;
+    niels_t ni;
+
+    curve448_scalar_t scalar1x;
```
and here

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+    curve448_scalar_halve(scalar1x, scalar1x);
+
+    for (i = s - 1; i >= 0; i--) {
+        if (i != (int)s - 1)
```
It us unfortunate to have this type cast here.
Would it be better to make s an int ?

mattcaswell replied:
I actually refactored the code a little to make `s` an unsigned int to match all of the others in this function.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+            mask_t invert;
+
+            for (k = 0; k < t; k++) {
+                unsigned int bit = i + s * (k + j * t);
```
empty line missing

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+
+    /* The point is now on the twisted curve.  Move it to untwisted. */
+    gf x, y, z, t;
+    curve448_point_t q;
```
move these variable declarations two lines up.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+    enc2[EDDSA_448_PRIVATE_BYTES - 1] &= ~0x80;
+
+    succ = gf_deserialize(p->y, enc2, 1, 0);
+#if 0 == 0
```
what?

mattcaswell replied:
Removed this.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+
+    {
+        /* 4-isogeny 2xy/(y^2-ax^2), (y^2+ax^2)/(2-y^2-ax^2) */
+        gf a, b, c, d;
```
empty line.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+                                                      out[X_PUBLIC_BYTES],
+                                                      const curve448_point_t p)
+{
+    curve448_point_t q;
```
empty line.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+# define NUMTRAILINGZEROS	numtrailingzeros
+static uint32_t numtrailingzeros(uint32_t i)
+{
+    unsigned int tmp;
```
this should be uint32_t

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+        i = tmp;
+        num -= 2;
+    }
+    if ((i << 1) != 0)
```
this should follow the same pattern (just in case the << is done in a integer promoted type):
`tmp = i << 1;`
`if (tmp != 0)`

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+            uint32_t odd = (uint32_t)current >> pos;
+            int32_t delta = odd & mask;
+
+            assert(position >= 0);
```
do not use assertion.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+            int32_t delta = odd & mask;
+
+            assert(position >= 0);
+            if (odd & 1 << (table_bits + 1))
```
this computes `odd & (1 << (table_bits + 1))` right?

mattcaswell replied:
Right - I actually had to go look this up to double check. Parens inserted to make it clearer.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+        }
+        current >>= 16;
+    }
+    assert(current == 0);
```
again, assert.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+
+    position++;
+    n = table_size - position;
+    for (i = 0; i < n; i++) {
```
unnecessary braces

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+                                               const curve448_scalar_t scalar2)
+{
+    const int table_bits_var = C448_WNAF_VAR_TABLE_BITS,
+        table_bits_pre = C448_WNAF_FIXED_TABLE_BITS;
```
just make a new declaration statement instead.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+
+    for (i--; i >= 0; i--) {
+        int cv = (i == control_var[contv].power), cp =
+            (i == control_pre[contp].power);
```
I think this is against our programming style,
just make a new declaration `int cp = `

mattcaswell replied:
Is it? Fixed anyway.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+        point_double_internal(combo, combo, i && !(cv || cp));
+
+        if (cv) {
+            assert(control_var[contv].addend);
```
assertion.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+        if (cv) {
+            assert(control_var[contv].addend);
+
+            if (control_var[contv].addend > 0) {
```
unnecessary braces.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+        }
+
+        if (cp) {
+            assert(control_pre[contp].addend);
```
assertion

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+        if (cp) {
+            assert(control_pre[contp].addend);
+
+            if (control_pre[contp].addend > 0) {
```
braces.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,747 @@
+    OPENSSL_cleanse(control_pre, sizeof(control_pre));
+    OPENSSL_cleanse(precmp_var, sizeof(precmp_var));
+
+    assert(contv == ncb_var);
```
assertion.

bernd-edlinger commented:
Sorry for so many commments...
I am gonna take a break now :)

richsalz commented:
Perhaps it is time to just remove all of the define tricks and write straightforward C code.  The only reason I can think to keep it is if we want to cherry-pick improvements that Mike makes to his code.  That seems unlikely.

bitwiseshiftleft commented:
32-byte alignment is required for avx2 operations.  On x86 at least, the library may be using avx2 even if the caller isn’t compiled with -mavx2.  We could instead change it so all the vector ops assume unaligned.

Sent from my phone.  Please excuse brevity and typos.

> On Feb 2, 2018, at 08:27, Bernd Edlinger <notifications@github.com> wrote:
> 
> @bernd-edlinger commented on this pull request.
> 
> In crypto/ec/curve448/field.h:
> 
> > +#ifndef __GF_H__
> +# define __GF_H__
> +
> +# include "internal/constant_time_locl.h"
> +# include <string.h>
> +# include <assert.h>
> +# include "word.h"
> +
> +# define NLIMBS (64/sizeof(word_t))
> +# define X_SER_BYTES 56
> +# define SER_BYTES 56
> +
> +# if defined(__GNUC__) || defined(__clang__)
> +#  define INLINE_UNUSED __inline__ __attribute__((unused,always_inline))
> +#  define RESTRICT __restrict__
> +#  define ALIGNED __attribute__((aligned(32)))
> this requests 32 byte alignment. The alignment is not measured in bits here.
> Was that the intention?
> That is more aligned that what we usually get from malloc.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

vdukhovni commented:
> @richsalz Perhaps it is time to just remove all of the define tricks and write straightforward C code. The only reason I can think to keep it is if we want to cherry-pick improvements that Mike makes to his code. That seems unlikely.

I disagree.  For highly-repetitive blocks of code with only minor parametric changes between repetitions, it becomes much more difficult to spot subtle errors by visually comparing the blocks than to just check the correctness of the `#define` and the parameters to each macro call. Sometimes readability of the correctness of the algorithm trumps readability of C control flow. If we had a real meta-compiler that could generate the C code from a higher level algorithm description we'd go with that, but the C-preprocessor is here in a way a poor-man's version of such a compiler.

mattcaswell commented:
I pushed a couple of commits to address the latest round of comments. I have either responded to all your comments above - or if not they were just style issues that are fixed.

bernd-edlinger commented on crypto/ec/curve448/ed448.h:
```diff
@@ -0,0 +1,198 @@
+ *
+ * For Ed25519, it is unsafe to use the same key for both prehashed and
+ * non-prehashed messages, at least without some very careful protocol-level
+ * disambiguation.  For Ed448 it is safe.  The C++ wrapper is designed to make
```
another reference to the C++ wrapper?

mattcaswell replied:
I fixed this and all other instances I could find.

bernd-edlinger commented on crypto/ec/curve448/ed448.h:
```diff
@@ -0,0 +1,198 @@
+ *
+ * For Ed25519, it is unsafe to use the same key for both prehashed and
+ * non-prehashed messages, at least without some very careful protocol-level
+ * disambiguation.  For Ed448 it is safe.  The C++ wrapper is designed to make
```
remove this sentence as well.

bernd-edlinger commented on crypto/ec/curve448/eddsa.c:
```diff
@@ -0,0 +1,359 @@
+               const uint8_t public_key[57], const uint8_t private_key[57],
+               const uint8_t *context, size_t context_len)
+{
+
```
remove empty line here.

bernd-edlinger commented on crypto/ec/curve448/field.h:
```diff
@@ -0,0 +1,172 @@
+# if defined(__GNUC__) || defined(__clang__)
+#  define INLINE_UNUSED __inline__ __attribute__((__unused__,__always_inline__))
+#  define RESTRICT __restrict__
+#  define ALIGNED __attribute__((aligned(32)))
```
underscores could be used for `__aligned__` as well.

bernd-edlinger replied:
Note however that based on the comment from @bitwiseshiftleft
the alignment here is used to match the alignment of a vector type
that is aliased with the gf_s struct at another place.
That is ok (from my GCC perspective), if the aliased vector is
a vector of word_t, thus the so called "alias set" of word_t and
vector of word_t is identical.
It could of course be more portable by placing all possible aliases
in a union inside gf_s, then the alignment would be automatically
correct.

mattcaswell replied:
I changed to use underscores. However I just removed the vector type code anyway (see below), so this alignment might be excessive. I've left it for now in case we resurrect the vector code later.

bernd-edlinger commented on crypto/ec/curve448/field.h:
```diff
@@ -0,0 +1,172 @@
+    for (i = 0; i < NLIMBS; i++) {
+#if ARCH_WORD_BITS == 32
+        constant_time_cond_swap_32((uint32_t)swap, (uint32_t *)&(x[0].limb[i]),
+                                   (uint32_t *)&(y->limb[i]));
```
are the type casts here really necessary?

mattcaswell replied:
Probably not. It was more a readability thing - but maybe it doesn't help that much. I removed them.

bernd-edlinger commented on crypto/ec/curve448/point_448.h:
```diff
@@ -0,0 +1,301 @@
+                         const curve448_scalar_t a, const curve448_scalar_t b);
+
+/*
+ * Multiply two scalars. The scalars may use the same memory.
```
Does this comment mean a and b may alias each other?
Or may a and out alias as well?

mattcaswell replied:
My understanding from looking at the code is that |a|, |b| and |out| may all be aliases. I updated the wording for all of these instances.

bernd-edlinger commented on crypto/ec/curve448/point_448.h:
```diff
@@ -0,0 +1,301 @@
+void curve448_scalar_halve(curve448_scalar_t out, const curve448_scalar_t a);
+
+/*
+ * Copy a scalar.  The scalars may use the same memory, in which case this
```
I think instead of "use the same memory", we should say "may alias".

bernd-edlinger replied:
So, "The scalars may alias each other, in which case...".

bernd-edlinger replied:
Please rewrite this comment.

bernd-edlinger replied:
Apparently this one slipped through.
I still believe the wording "may use the same memory" is awkward,
the correct term is something like "The scalars may alias each other"

mattcaswell replied:
Darn - I thought I did this. Fixed now.

bernd-edlinger commented on crypto/ec/curve448/scalar.c:
```diff
@@ -0,0 +1,234 @@
+{
+    c448_word_t mask = 0 - (a->limb[0] & 1);
+    c448_dword_t chain = 0;
+    unsigned int i;
```
empty line here.

bernd-edlinger commented on crypto/ec/curve448/arch_32/f_impl.h:
```diff
@@ -0,0 +1,62 @@
+# define HEADER_ARCH_32_F_IMPL_H
+
+# define GF_HEADROOM 2
+# define LIMB(x) (x)&((1<<28)-1), (x)>>28
```
please add parentheses around both expressions to make this look like the SC_LIMB macro

bernd-edlinger commented on crypto/ec/curve448/field.h:
```diff
@@ -0,0 +1,172 @@
+/* Constant time, if (neg) x=-x; */
+static ossl_inline void gf_cond_neg(gf x, mask_t neg)
+{
+    gf y;
```
empty line here.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,726 @@
+                                       1 << (t - 1), tab);
+
+            cond_neg_niels(ni, invert);
+            if ((i != s) || j != 0) {
```
unnecessary braces here.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,726 @@
+            for (k = 0; k < t; k++) {
+                unsigned int bit = (i - 1) + s * (k + j * t);
+
+                if (bit < C448_SCALAR_BITS) {
```
unnecessary braces here.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,726 @@
+
+    if (i < 0) {
+        curve448_point_copy(combo, curve448_point_identity);
+        return;
```
Don't we hate return before else ?

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,726 @@
+        gf_cond_swap(z2, z3, swap);
+        swap = k_t;
+
+        gf_add_nr(t1, x2, z2);  /* A = x2 + z2 *//* 2+e */
```
what does this comment `/* 2+e */` mean?

mattcaswell replied:
I don't know.

bernd-edlinger replied:
Maybe place a comment here with what @bitwiseshiftleft said:

/* The _nr skips coefficient reduction. In the following comments,
 "2+e"  is saying that the coefficients are at most 2+epsilon times the reduction limit. */

bernd-edlinger replied:
and, maybe, add a blank between the two comments?
I mean `*//*` this looks almost like a C++ comment inside a C comment...

mattcaswell replied:
Done. The original code used to have the blank between the two comments, but `openssl-format-source` screwed it up.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,726 @@
+    curve448_scalar_decode_long(the_scalar, scalar2, sizeof(scalar2));
+
+    /* Compensate for the encoding ratio */
+    for (i = 1; i < X448_ENCODE_RATIO; i <<= 1) {
```
unnecessary braces.

bernd-edlinger commented on crypto/ec/curve448/scalar.c:
```diff
@@ -0,0 +1,234 @@
+{
+    unsigned int i, j, k = 0;
+
+    for (i = 0; i < C448_SCALAR_LIMBS; i++) {
```
does the nested for count as one or two statements?
if one, then the braces should be removed.

mattcaswell replied:
I think two. I left the braces.

bernd-edlinger replied:
ok.

bernd-edlinger commented on crypto/ec/curve448/eddsa.c:
```diff
@@ -0,0 +1,359 @@
+    const char *dom_s = "SigEd448";
+    uint8_t dom[2];
+
+    dom[0] = 2 + word_is_zero(prehashed) + word_is_zero(for_prehash);
```
It is a bit surprising to use constant_time_is_zero_32 here.
Does this need the constant time hammer at all?
If yes, at least use constant_time_is_zero_8?

mattcaswell replied:
None of the rest of this code is constant time, so I don't think it is necessary. I removed it.

bernd-edlinger commented on crypto/ec/curve448/eddsa.c:
```diff
@@ -0,0 +1,359 @@
+    dom[0] = 2 + word_is_zero(prehashed) + word_is_zero(for_prehash);
+    dom[1] = (uint8_t)context_len;
+
+    if (context_len > UINT8_MAX)
```
maybe move this before the type cast above?

bernd-edlinger commented on crypto/ec/curve448/field.h:
```diff
@@ -0,0 +1,172 @@
+/* Square x, n times. */
+static ossl_inline void gf_sqrn(gf_s * RESTRICT y, const gf x, int n)
+{
+    gf tmp;
```
empty line here.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,726 @@
+{
+    mask_t ret;
+
+    gf t1, t2;
```
move this one line up.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,726 @@
+        if (w < (C448_SCALAR_BITS - 1) / 16 + 1) {
+            /* Refill the 16 high bits of current */
+            current += (uint32_t)((scalar->limb[w / B_OVER_16]
+                       >> (16 * (w %  B_OVER_16))) << 16);
```
I see two blanks between % and B_OVER_16, one would look better.

bernd-edlinger commented on crypto/ec/curve448/field.h:
```diff
@@ -0,0 +1,172 @@
+
+# include "f_impl.h"            /* Bring in the inline implementations */
+
+# ifndef LIMBPERM
```
why is this ifndef needed?

mattcaswell replied:
It doesn't seem to be. I removed it.

bernd-edlinger commented on crypto/ec/curve448/word.h:
```diff
@@ -0,0 +1,159 @@
+
+# if defined(__AVX2__)
+#  define VECTOR_ALIGNED __attribute__((aligned(32)))
+typedef uint64x4_t uint64xn_t;
```
I don't see where any of theses vector types are used.

mattcaswell replied:
They used to be used in the platform specific optimisation files that I removed earlier in the review process. I've removed these remnants of that code too now.

bernd-edlinger commented on crypto/ec/curve448/word.h:
```diff
@@ -0,0 +1,159 @@
+typedef uint64x4_t uint64xn_t;
+typedef uint32x8_t uint32xn_t;
+# elif defined(__SSE2__)
+#  define VECTOR_ALIGNED __attribute__((aligned(16)))
```
I don't really see what VECTOR_ALIGNED is good for.
It seems that it is always used with things that contain gf objects
which are already 32-byte aligned (already use ALIGNED).

mattcaswell replied:
Since I removed the vector code, I removed this too.

mattcaswell commented:
New commits pushed addressing all the latest feedback. I either responded to your feedback or just fixed the style issues.

bernd-edlinger commented on crypto/ec/curve448/curve448_tables.c:
```diff
@@ -0,0 +1,477 @@
+        }}
+    }
+};
+extern const struct curve448_precomputed_s *curve448_precomputed_base;
```
Hmm.  This could mean that a header file is missing?

bernd-edlinger replied:
`#include <point_448.h>` here or something?

mattcaswell replied:
Should be fixed now.

bernd-edlinger commented on crypto/ec/curve448/eddsa.c:
```diff
@@ -0,0 +1,359 @@
+#include "curve448_lcl.h"
+#include "word.h"
+#include "ed448.h"
+#include <string.h>
```
I'd like to move this system header file a few places up.

mattcaswell replied:
Done. I moved it to the top so we have system headers first; public openssl headers next; private openssl headers last.

bernd-edlinger commented on crypto/ec/curve448/eddsa.c:
```diff
@@ -0,0 +1,359 @@
+
+static void clamp(uint8_t secret_scalar_ser[EDDSA_448_PRIVATE_BYTES])
+{
+    uint8_t hibit = (1 << 0) >> 1;
```
hey, what is this `(1 << 0) >> 1` ?

mattcaswell replied:
It's a remnant of the original code approach. We've just imported the curve448 stuff, but the original codebase supported 25519 as well. Some of the code was templated and reused between the two implementations. I removed all of that as part of this work. This code originally looked like this:

````
    uint8_t hibit = (1<<$(gf_bits % 8))>>1;
````
I have now simplified this function accordingly.

mattcaswell commented:
Some fixup commits pushed to address some travis/appveyor issues and your feedback as well as a commit to address your question about the clamp function.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,721 @@
+#define C448_WNAF_FIXED_TABLE_BITS 5
+#define C448_WNAF_VAR_TABLE_BITS 3
+
+static const int EDWARDS_D = -39081;
```
why is this not simply a define?

mattcaswell replied:
It is now.,

bernd-edlinger commented on crypto/ec/curve448/scalar.c:
```diff
@@ -0,0 +1,235 @@
+    unsigned int i;
+    c448_word_t borrow;
+
+    for (i = 0; i < C448_SCALAR_LIMBS; i++) {
```
would our FOR_LIMB macro make sense here?

mattcaswell replied:
Yes. I changed it here and other similar places in this file.

bernd-edlinger commented on crypto/ec/curve448/eddsa.c:
```diff
@@ -0,0 +1,350 @@
+        return C448_FAILURE;
+
+    dom[0] = 2 + (uint8_t)(prehashed == 0 ? -1 : 0)
+             + (uint8_t)(for_prehash == 0 ?  -1 : 0);
```
maybe it would be better this way: `(uint8_t)(2 - (prehashed == 0 ? 1 : 0) - ...)`

mattcaswell replied:
Ok. Done.

bernd-edlinger commented on crypto/ec/curve448/f_generic.c:
```diff
@@ -0,0 +1,205 @@
+
+    gf_add(y, x, x);
+    gf_strong_reduce(y);
+    return 0 - (y->limb[0] & 1);
```
would using "word_is_zero" make this more constant time?

mattcaswell replied:
This is testing whether the lsb of the word is set - which is not what "word_is_zero" does. I suppose you could do it like this:
````
    return ~word_is_zero(y->limb[0] & 1);
````
But its not clear to me that that is any better from a readability point of view. As for "more" constant time - I'm not sure why you think this isn't already constant time?

bernd-edlinger replied:
I was not sure if this could be optimized to y-> limb[0] & 1 ? 0 : -1
but I don't think that is likely to happen.

mattcaswell replied:
Let's hope not. Ultimately we are the mercy of the compiler for what it decides to do. We can only do so much.

bernd-edlinger commented on crypto/ec/curve448/word.h:
```diff
@@ -0,0 +1,99 @@
+# include "arch_intrinsics.h"
+
+# include "curve448utils.h"
+# include <stdlib.h>
```
maybe move this system include up as well.

mattcaswell replied:
Done

bernd-edlinger commented on crypto/ec/curve448/word.h:
```diff
@@ -0,0 +1,99 @@
+# if C448_WORD_BITS == 64
+#  define SC_LIMB(x) (x)
+# elif C448_WORD_BITS == 32
+#  define SC_LIMB(x) ((uint32_t)x),(x >> 32)
```
macro parameter 'x' is not in parentheses

mattcaswell replied:
Done

bitwiseshiftleft commented:
The reason for ignore_result() was that casting to void isn’t sufficient to suppress __attribute__((warn_unused_result)) in some compilers.  If you’re removing that attribute too, then you don’t need ignore_result().


> On Feb 6, 2018, at 11:43 PM, Bernd Edlinger <notifications@github.com> wrote:
> 
> @bernd-edlinger commented on this pull request.
> 
> In crypto/ec/curve448/word.h <https://github.com/openssl/openssl/pull/5105#discussion_r166538145>:
> 
> > +    /* On most arches this will be optimized to a simple cast. */
> +    mask_t ret = 0;
> +    unsigned int i;
> +    unsigned int limit = sizeof(c448_bool_t) / sizeof(mask_t);
> +
> +    if (limit < 1)
> +        limit = 1;
> +    for (i = 0; i < limit; i++)
> +        ret |= ~word_is_zero(m >> (i * 8 * sizeof(word_t)));
> +
> +    return ret;
> +}
> +
> +static ossl_inline void ignore_result(c448_bool_t boo)
> +{
> +    (void)boo;
> yes, agreed. I was just unsure if this was supposed to force the compiler to not optimize the
> ignored expression away.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/openssl/openssl/pull/5105#discussion_r166538145>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AARlXLzcuOnGqq6o7C_0hHHf65LWKbxLks5tSVQtgaJpZM4Ri79d>.
>

bitwiseshiftleft commented:
The _nr skips coefficient reduction.  /* 2+e */ is saying that the coefficients are at most 2+epsilon times the reduction limit.  An older version of the code used _nr more frequently and was littered with this sort of comment, but the current version uses it only in one or two functions to make checking easier.

> On Feb 7, 2018, at 12:50 AM, Bernd Edlinger <notifications@github.com> wrote:
> 
> @bernd-edlinger commented on this pull request.
> 
> In crypto/ec/curve448/curve448.c <https://github.com/openssl/openssl/pull/5105#discussion_r166550317>:
> 
> > +
> +        /* Scalar conditioning */
> +        if (t / 8 == 0)
> +            sb &= -(uint8_t)COFACTOR;
> +        else if (t == X_PRIVATE_BITS - 1)
> +            sb = -1;
> +
> +        k_t = (sb >> (t % 8)) & 1;
> +        k_t = 0 - k_t;             /* set to all 0s or all 1s */
> +
> +        swap ^= k_t;
> +        gf_cond_swap(x2, x3, swap);
> +        gf_cond_swap(z2, z3, swap);
> +        swap = k_t;
> +
> +        gf_add_nr(t1, x2, z2);  /* A = x2 + z2 *//* 2+e */
> what does this comment /* 2+e */ mean?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/openssl/openssl/pull/5105#pullrequestreview-94628160>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AARlXP6IUlpx3cwJmDLErk2qknp8pmYjks5tSWPGgaJpZM4Ri79d>.
>

bitwiseshiftleft commented:
Could use -! instead.

> On Feb 7, 2018, at 1:56 AM, Bernd Edlinger <notifications@github.com> wrote:
> 
> @bernd-edlinger commented on this pull request.
> 
> In crypto/ec/curve448/eddsa.c <https://github.com/openssl/openssl/pull/5105#discussion_r166567439>:
> 
> > +        secret_scalar_ser[EDDSA_448_PRIVATE_BYTES - 2] |= 0x80;
> +    } else {
> +        secret_scalar_ser[EDDSA_448_PRIVATE_BYTES - 1] &= hibit - 1;
> +        secret_scalar_ser[EDDSA_448_PRIVATE_BYTES - 1] |= hibit;
> +    }
> +}
> +
> +static c448_error_t hash_init_with_dom(EVP_MD_CTX *hashctx, uint8_t prehashed,
> +                                       uint8_t for_prehash,
> +                                       const uint8_t *context,
> +                                       size_t context_len)
> +{
> +    const char *dom_s = "SigEd448";
> +    uint8_t dom[2];
> +
> +    dom[0] = 2 + word_is_zero(prehashed) + word_is_zero(for_prehash);
> It is a bit surprising to use constant_time_is_zero_32 here.
> Does this need the constant time hammer at all?
> If yes, at least use constant_time_is_zero_8?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/openssl/openssl/pull/5105#pullrequestreview-94647994>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AARlXAwKQ9GSuJ9XCVgWuYDuKLzUnSKRks5tSXNpgaJpZM4Ri79d>.
>

bitwiseshiftleft commented:
The NEON version of the code permutes the limbs within a gf for efficiency reasons.  The other arches don’t.

> On Feb 7, 2018, at 2:22 AM, Bernd Edlinger <notifications@github.com> wrote:
> 
> @bernd-edlinger commented on this pull request.
> 
> In crypto/ec/curve448/field.h <https://github.com/openssl/openssl/pull/5105#discussion_r166573958>:
> 
> > +void gf_sub(gf out, const gf a, const gf b);
> +void gf_mul(gf_s * RESTRICT out, const gf a, const gf b);
> +void gf_mulw_unsigned(gf_s * RESTRICT out, const gf a, uint32_t b);
> +void gf_sqr(gf_s * RESTRICT out, const gf a);
> +mask_t gf_isr(gf a, const gf x); /** a^2 x = 1, QNR, or 0 if x=0.  Return true if successful */
> +mask_t gf_eq(const gf x, const gf y);
> +mask_t gf_lobit(const gf x);
> +mask_t gf_hibit(const gf x);
> +
> +void gf_serialize(uint8_t *serial, const gf x, int with_highbit);
> +mask_t gf_deserialize(gf x, const uint8_t serial[SER_BYTES], int with_hibit,
> +                      uint8_t hi_nmask);
> +
> +# include "f_impl.h"            /* Bring in the inline implementations */
> +
> +# ifndef LIMBPERM
> why is this ifndef needed?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/openssl/openssl/pull/5105#pullrequestreview-94655990>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AARlXOICSDisdYWrZQJj0kll1YOVqIMHks5tSXlogaJpZM4Ri79d>.
>

mattcaswell commented:
> The NEON version of the code permutes the limbs within a gf for efficiency reasons.  The other arches don’t.

Right. We're only using the 32-bit reference version of the code in this PR so we don't have the architecture specific optimisations here.

mattcaswell commented:
More fixes pushed as described in my response to your comments.

bernd-edlinger commented:
> The reason for ignore_result() was that casting to void isn’t sufficient to suppress __attribute__((warn_unused_result)) in some compilers. If you’re removing that attribute too, then you don’t need ignore_result().

Hmmm, well, maybe this attribute was not really helpful then.
Should we remove it?

mattcaswell commented:
> Hmmm, well, maybe this attribute was not really helpful then.
Should we remove it?

Well it seems we only use it in two places anyway. And in both of those cases the functions in question do not have that attribute against them anyway. I'll remove it - although I will still cast the result to void to make it clear that ignoring the result is intentional.

bernd-edlinger commented on crypto/ec/curve448/curve448.c:
```diff
@@ -0,0 +1,722 @@
+    }
+};
+
+#define TWISTED_D ((EDWARDS_D)-1)
```
Matt, whenever I looked at "(EDWARDS_D)-1" I thought it was a type cast.
But now I see, it is not...
Please write this as "EDWARDS_D - 1".

mattcaswell replied:
Done

mattcaswell commented:
Travis failures appear unrelated. This is the afalg engine issue that is in master at the moment.

mattcaswell commented:
Latest fixes pushed.

bernd-edlinger commented on crypto/ec/curve448/eddsa.c:
```diff
@@ -0,0 +1,350 @@
+        return C448_FAILURE;
+
+    dom[0] = 2 - (uint8_t)(prehashed == 0 ? 1 : 0)
+             - (uint8_t)(for_prehash == 0 ? 1 : 0);
```
This is okay, but actually I meant to dothe uint8_t-cast on the whole expression,
because the arithmetic is always done with integer-promoted values.

bernd-edlinger commented on crypto/ec/curve448/scalar.c:
```diff
@@ -0,0 +1,238 @@
+#define WBITS C448_WORD_BITS   /* NB this may be different from ARCH_WORD_BITS */
+
+const curve448_scalar_t curve448_scalar_one = {{{1}}};
+const curve448_scalar_t  curve448_scalar_zero = {{{0}}};
```
this line does not align with the one above.
Please remove the extra space before "curve448"

dot-asm commented on crypto/ec/curve448/curve448utils.h:
```diff
@@ -0,0 +1,107 @@
+            break; \
+        } while(1); \
+    } while(0)
+#endif
```
This implies that compiler is capable of recognizing that this is so to say "constant expression" and not generate things like dispatch table to be used with indirect jump or long chain of conditional branches. But here is a question. If you trust compiler with that, then how come you don't trust compiler to unroll the small[!] loop with constant controlling expression? I mean compilers *are* programmed to make this exact kind of trade-off decision. Besides, it's not actually obvious that unrolled loops are always good for performance. Well, they are likely perform better in benchmark, but in real life by the time you execute piece of code it's more than likely to be evicted from cache and it surely takes longer time to bring in 20KB than 4KB, just to execute it once prior it's evicted till next time around...  BTW, I do observe indirect jumps in code generated with GCC...

dot-asm replied:
> What would you expect to be generated here?

I would expect that switch(8) is compiled with no branches whatsoever. But if I can't expect that, and apparently we can't, then trade-off makes no sense. I mean loop would do just as fine. As already implied, there is no reason to feel bad about loops. On high-end processors loop control instructions would cost virtually nothing (modulo occasionally mispredicted branches). Loops can actually be more efficient on most popular platform (a.k.a. Intel). This is because it caches *decoded* [and fused] instructions.

bernd-edlinger replied:
When you have a literal in the switch that should of course be easy to optimize the jump away.
Do you have a test case where the code generation is sub-optimal?

dot-asm replied:
Let's reconstruct events a little bit. Me: "I'm skeptical about this even with 'constant expressions', and I do note *some* indirect jumps". You: "It's fancy term thing, I would expect indirect jump". Me: "Oh! Looks like we are bound to observe indirect jump, then I do suggest we omit it." To summarize. I didn't actually performed deep analysis, so I don't really have test case. And I'd be reluctant to search for one, because it can be time consuming. Also, I view it of somewhat lesser relevance what particular compiler version does, current or future. [Do note that I'm not saying "*no* relevance", but "*lesser* relevance" :-)] More relevant question [in my mind] is if this (or any particular) piece of code worth the effort of maintaining across *broad* scale of compilers, i.e. without making assumption about specific one or its version. And as already said I'd argue that it's appropriate to rely on compiler to make the trade-off decision about small loop unroll. Because it's commonly recognized optimization technique.

bernd-edlinger replied:
Yes, good point.

bitwiseshiftleft commented:
I would recommend against Duff’s device.  It’s a hard-to-read hack which isn’t especially useful on a modern compiler anyway.

> On Feb 7, 2018, at 10:47 PM, Bernd Edlinger <notifications@github.com> wrote:
> 
> @bernd-edlinger commented on this pull request.
> 
> In crypto/ec/curve448/curve448utils.h <https://github.com/openssl/openssl/pull/5105#discussion_r166844252>:
> 
> > +            switch((_end) - (_i)) { \
> +            default: \
> +            case 8: {_x}; (_i)++; /* fall through */ \
> +            case 7: {_x}; (_i)++; /* fall through */ \
> +            case 6: {_x}; (_i)++; /* fall through */ \
> +            case 5: {_x}; (_i)++; /* fall through */ \
> +            case 4: {_x}; (_i)++; /* fall through */ \
> +            case 3: {_x}; (_i)++; /* fall through */ \
> +            case 2: {_x}; (_i)++; /* fall through */ \
> +            case 1: {_x}; (_i)++; continue; \
> +            case 0: break; \
> +            } \
> +            break; \
> +        } while(1); \
> +    } while(0)
> +#endif
> This is an old-school technique called a Duff's Device.
> I would expect to see a compare against 9 and conditional branch to
> the default label, followed by an indirect branch via call table here.
> What would you expect to be generated here?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/openssl/openssl/pull/5105#discussion_r166844252>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AARlXBNpWWqWu9A92H3Eun_beX8kCh19ks5tSpiTgaJpZM4Ri79d>.
>

bernd-edlinger commented:
> I would recommend against Duff’s device.

That's ok for me, then we should rip out the FOR_LIMB stuff.

mattcaswell commented:
I rebased this, fixed a travis failure (complaining about an unused result from curve448_scalar_decode), resolved the outstanding nits above, and removed all of the FOR_LIMB stuff (replacing all those commits with a single one which removes it all).

bernd-edlinger commented on crypto/ec/curve448/arch_32/f_impl.c:
```diff
@@ -0,0 +1,95 @@
+    uint32_t mask = (1 << 28) - 1;
+    int i;
+
+    assert(b < 1 << 28);
```
maybe write the condition using `mask` as `b <= mask` ?

bernd-edlinger commented on crypto/ec/curve448/arch_32/f_impl.h:
```diff
@@ -0,0 +1,60 @@
+    uint32_t co1 = ((1 << 28) - 1) * amt, co2 = co1 - amt;
+
+    for (i = 0; i < NLIMBS; i++)
+        a->limb[i] += (i == sizeof(*a) / sizeof(a->limb[0]) / 2) ? co2 : co1;
```
I wonder if `sizeof(*a) / sizeof(a->limb[0])` could be replaced with `NLIMBS` ?
Thus making the condition something like `(i == NLIMBS / 2)` ?

bernd-edlinger commented on crypto/ec/curve448/arch_32/f_impl.h:
```diff
@@ -0,0 +1,60 @@
+void gf_weak_reduce(gf a)
+{
+    uint32_t mask = (1 << 28) - 1;
+    uint32_t tmp = a->limb[15] >> 28;
```
is this by chance `NLIMBS-1` ?

bernd-edlinger commented on crypto/ec/curve448/arch_32/f_impl.h:
```diff
@@ -0,0 +1,60 @@
+    uint32_t tmp = a->limb[15] >> 28;
+    unsigned int i;
+
+    a->limb[8] += tmp;
```
is 8 by chance `NLIMBS/2` ?

mattcaswell commented:
All issues above resolved, and a new commit pushed.

bernd-edlinger commented on crypto/ec/curve448/arch_32/f_impl.h:
```diff
@@ -0,0 +1,60 @@
+    unsigned int i;
+
+    a->limb[NLIMBS / 2] += tmp;
+    for (i = 15; i > 0; i--)
```
`i = NLIMBS - 1` right?

mattcaswell replied:
Right. Fixup commit pushed addressing this.

bernd-edlinger commented on crypto/ec/curve448/word.h:
```diff
@@ -0,0 +1,92 @@
+# include "arch_intrinsics.h"
+# include "curve448utils.h"
+
+# if defined(__ARM_NEON__)
```
are these include files still necessary?

bernd-edlinger replied:
I mean the ones below.

mattcaswell replied:
No - not needed. I've removed them.

mattcaswell commented:
New commits pushed addressing latest comments.

kroeckx commented on include/internal/constant_time_locl.h:
```diff
@@ -82,6 +87,11 @@ static ossl_inline unsigned int constant_time_select(unsigned int mask,
+
+/* Convenience method for uint32_t. */
+static ossl_inline uint32_t constant_time_select_32(uint32_t mask, uint32_t a,
+                                                    uint32_t b);
```
Please update the test suite

mattcaswell commented:
Constant time tests updated as requested by @kroeckx.

@bernd-edlinger, please reconfirm!

bernd-edlinger commented:
appappveyor is relevant:
```
..\test\constant_time_test.c(105): warning C4244: 'function': conversion from 'uint64_t' to 'const unsigned int', possible loss of data
..\test\constant_time_test.c(141): warning C4244: 'function': conversion from 'uint64_t' to 'uint32_t', possible loss of data
```

travis as well:
```
test/constant_time_test.d.tmp -MT test/constant_time_test.o -c -o test/constant_time_test.o test/constant_time_test.c
test/constant_time_test.c: In function 'test_32values':
test/constant_time_test.c:349:18: error: iteration 9ul invokes undefined behavior [-Werror=aggressive-loop-optimizations]
         uint32_t f = test_values_32[j];
                  ^
test/constant_time_test.c:348:5: note: containing loop
     for (j = 0; j < sizeof(test_values_32); ++j) {
     ^
```

mattcaswell commented:
Fixup commit pushed to address the Travis and Appveyor problems. Lets see if they are happy...

bernd-edlinger commented:
Travis is still not working:

```
       # ERROR: (bool) 'constant_time_is_zero_32(a) == CONSTTIME_FALSE_32 == true' failed @ ../_srcdist/test/constant_time_test.c:104
        # false
        not ok 10 - iteration 10
    not ok 2 - test_binops
```

bernd-edlinger commented on test/constant_time_test.c:
```diff
@@ -221,56 +246,60 @@ static int test_sizeofs(void)
 {
     unsigned int a = test_values[i];
-    unsigned int g = test_values_s[i];
+    uint32_t a32 = test_values_32[i];
```
test_values_32 has only 9 elements,
while test_values has 10.

bernd-edlinger replied:
ADD_ALL_TESTS(test_binops, OSSL_NELEM(test_values));

mattcaswell replied:
Yeah - I figured that out. New commit pushed.

mattcaswell commented:
CIs seem happy now.

mattcaswell commented:
Pushed!!! Thanks @bernd-edlinger for all your work reviewing this!
